File: src/services/artifactObjectiveService.ts
// Artifact Objective Service - TypeScript implementation based on artifact_objective_service.py
// Handles artifact objective parsing and formatting

export interface NameLink {
  name: string;
  link: string | null;
}

export type FetchLookup = (id: string) => Promise<NameLink | null>;

const TASK_PERK_RE = /task_perk\d+_/i;

function coerceInt(value: any): number | null {
  try {
    return parseInt(String(value));
  } catch (error) {
    return null;
  }
}

function extractZoneId(poiTag: string): string | null {
  const token = (poiTag || "").trim();
  if (!token) return null;
  
  const last = token.split("_").pop();
  if (last && /^\d+$/.test(last)) {
    return last;
  }
  
  const match = token.match(/(\d+)/);
  return match ? match[1] : null;
}

function prettifyIdentifier(identifier: string): string {
  const raw = (identifier || "").replace(/_/g, " ").trim();
  if (!raw) return "";
  
  const spaced = raw.replace(/(?<=\w)([A-Z])/g, " $1");
  const tokens = spaced.split(/\s+/);
  if (!tokens.length) return spaced;
  
  const norm = (token: string) => 
    /^[A-Z]+$/.test(token) ? token : token.charAt(0).toUpperCase() + token.slice(1).toLowerCase();
  
  return tokens.map(norm).join(" ");
}

function fallbackLocationLabel(identifier: string): string | null {
  const token = (identifier || "").trim();
  if (!token) return null;
  
  let cleaned = token
    .replace(/^(Dungeon|Activity)/, "")
    .replace(/\d+$/, "");
  
  cleaned = cleaned.replace(/(?<!^)(?=[A-Z])/g, " ").trim();
  return cleaned || null;
}

async function formatKillContribution(
  entry: any,
  fetchCreature: FetchLookup,
  fetchZone: FetchLookup,
  fetchGamemode: FetchLookup
): Promise<string | null> {
  const qty = coerceInt(entry.TargetQty);
  const killEnemy = String(entry.KillEnemyType || "").trim();
  
  let targetName: string | null = null;
  let targetLink: string | null = null;
  
  if (killEnemy) {
    const lookup = await fetchCreature(killEnemy);
    if (lookup) {
      targetName = lookup.name;
      targetLink = lookup.link;
    } else {
      targetName = prettifyIdentifier(killEnemy);
    }
  }
  
  if (!targetName) return null;
  
  const target = targetLink ? `[${targetName}](${targetLink})` : targetName;
  
  const parts = ["Defeat"];
  if (qty && qty > 1) {
    parts.push(String(qty));
  }
  parts.push(target);
  
  let zoneId = extractZoneId(String(entry.POITag || ""));
  if (!zoneId) {
    const territory = coerceInt(entry.TerritoryID);
    if (territory && territory > 0) {
      zoneId = String(territory);
    }
  }
  
  let location: string | null = null;
  let preposition = "at";
  
  if (zoneId) {
    const zoneLookup = await fetchZone(zoneId);
    if (zoneLookup && zoneLookup.name) {
      location = zoneLookup.link ? `[${zoneLookup.name}](${zoneLookup.link})` : zoneLookup.name;
    }
  }
  
  if (!location) {
    const gamemodeId = String(entry.GameModeID || "").trim();
    if (gamemodeId) {
      const gmLookup = await fetchGamemode(gamemodeId);
      if (gmLookup && gmLookup.name) {
        preposition = "in";
        location = gmLookup.link ? `[${gmLookup.name}](${gmLookup.link})` : gmLookup.name;
      } else if (gmLookup === null) {
        const fallback = fallbackLocationLabel(gamemodeId);
        if (fallback) {
          preposition = "in";
          location = fallback;
        }
      }
    }
  }
  
  let text = parts.join(" ");
  if (location) {
    text += ` ${preposition} ${location}`;
  }
  
  return text;
}

function formatGameEvent(entry: any): string | null {
  const qty = coerceInt(entry.TargetQty);
  if (qty && qty > 1) {
    return `Capture ${qty} Forts`;
  }
  return "Capture the Fort";
}

export async function buildArtifactObjectives(
  itemId: string,
  tasks: any[],
  fetchCreature: FetchLookup,
  fetchZone: FetchLookup,
  fetchGamemode: FetchLookup
): Promise<string[]> {
  if (!itemId) return [];
  
  const iid = String(itemId).toLowerCase();
  const relevant: any[] = [];
  
  for (const entry of tasks || []) {
    if (!entry || typeof entry !== 'object') continue;
    
    const taskId = String(entry.TaskID || "");
    if (!taskId || !taskId.toLowerCase().includes(iid)) continue;
    if (!TASK_PERK_RE.test(taskId.toLowerCase())) continue;
    
    relevant.push(entry);
  }
  
  relevant.sort((a, b) => String(a.TaskID || "").localeCompare(String(b.TaskID || "")));
  
  const results: string[] = [];
  
  for (const entry of relevant) {
    const taskType = String(entry.Type || "").trim();
    let text: string | null = null;
    
    if (taskType === "TaskKillContribution") {
      text = await formatKillContribution(entry, fetchCreature, fetchZone, fetchGamemode);
    } else if (taskType === "TaskGameEvent") {
      text = formatGameEvent(entry);
    }
    
    if (text) {
      results.push(text);
    }
  }
  
  return results;
}

File: test/python_examples/artifact_objective_service.py
from __future__ import annotations

import re
from typing import Awaitable, Callable, Dict, Iterable, List, Optional, Tuple

NameLink = Tuple[str, Optional[str]]
FetchLookup = Callable[[str], Awaitable[Optional[NameLink]]]

_TASK_PERK_RE = re.compile(r"task_perk\d+_", re.IGNORECASE)


def _coerce_int(value: object) -> Optional[int]:
    try:
        iv = int(str(value))
    except (TypeError, ValueError):
        return None
    return iv


def _extract_zone_id(poi_tag: str) -> Optional[str]:
    token = (poi_tag or "").strip()
    if not token:
        return None
    last = token.split("_")[-1]
    if last.isdigit():
        return last
    match = re.search(r"(\d+)", token)
    return match.group(1) if match else None


def _prettify_identifier(identifier: str) -> str:
    raw = (identifier or "").replace("_", " ").strip()
    if not raw:
        return ""
    spaced = re.sub(r"(?<=\w)([A-Z])", r" \1", raw)
    tokens = spaced.split()
    if not tokens:
        return spaced
    def _norm(token: str) -> str:
        return token if token.isupper() else token.capitalize()
    return " ".join(_norm(t) for t in tokens)


def _fallback_location_label(identifier: str) -> Optional[str]:
    token = (identifier or "").strip()
    if not token:
        return None
    token = re.sub(r"^(Dungeon|Activity)", "", token)
    token = re.sub(r"\d+$", "", token)
    spaced = re.sub(r"(?<!^)(?=[A-Z])", " ", token).strip()
    return spaced or None


async def _format_kill_contribution(
    entry: Dict[str, object],
    fetch_creature: FetchLookup,
    fetch_zone: FetchLookup,
    fetch_gamemode: FetchLookup,
) -> Optional[str]:
    qty = _coerce_int(entry.get("TargetQty"))
    kill_enemy = str(entry.get("KillEnemyType") or "").strip()
    target_name = None
    target_link = None
    if kill_enemy:
        lookup = await fetch_creature(kill_enemy)
        if lookup:
            target_name, target_link = lookup
        else:
            target_name = _prettify_identifier(kill_enemy)
    if not target_name:
        return None
    target = f"[{target_name}]({target_link})" if target_link else target_name

    parts = ["Defeat"]
    if qty and qty > 1:
        parts.append(str(qty))
    parts.append(target)

    zone_id = _extract_zone_id(str(entry.get("POITag") or ""))
    if not zone_id:
        territory = _coerce_int(entry.get("TerritoryID"))
        if territory and territory > 0:
            zone_id = str(territory)

    location = None
    preposition = "at"
    if zone_id:
        zone_lookup = await fetch_zone(zone_id)
        if zone_lookup and zone_lookup[0]:
            z_name, z_link = zone_lookup
            location = f"[{z_name}]({z_link})" if z_link else z_name
    if not location:
        gamemode_id = str(entry.get("GameModeID") or "").strip()
        if gamemode_id:
            gm_lookup = await fetch_gamemode(gamemode_id)
            if gm_lookup and gm_lookup[0]:
                preposition = "in"
                g_name, g_link = gm_lookup
                location = f"[{g_name}]({g_link})" if g_link else g_name
            elif gm_lookup is None:
                fallback = _fallback_location_label(gamemode_id)
                if fallback:
                    preposition = "in"
                    location = fallback

    text = " ".join(parts)
    if location:
        text += f" {preposition} {location}"
    return text


def _format_game_event(entry: Dict[str, object]) -> Optional[str]:
    qty = _coerce_int(entry.get("TargetQty"))
    if qty and qty > 1:
        return f"Capture {qty} Forts"
    return "Capture the Fort"


async def build_artifact_objectives(
    item_id: str,
    tasks: Iterable[Dict[str, object]],
    fetch_creature: FetchLookup,
    fetch_zone: FetchLookup,
    fetch_gamemode: FetchLookup,
) -> List[str]:
    if not item_id:
        return []
    iid = str(item_id).lower()
    relevant: List[Dict[str, object]] = []
    for entry in tasks or []:
        if not isinstance(entry, dict):
            continue
        task_id = str(entry.get("TaskID") or "")
        if not task_id or iid not in task_id.lower():
            continue
        if not _TASK_PERK_RE.search(task_id.lower()):
            continue
        relevant.append(entry)
    relevant.sort(key=lambda e: str(e.get("TaskID") or ""))

    results: List[str] = []
    for entry in relevant:
        task_type = str(entry.get("Type") or "").strip()
        text: Optional[str]
        if task_type == "TaskKillContribution":
            text = await _format_kill_contribution(entry, fetch_creature, fetch_zone, fetch_gamemode)
        elif task_type == "TaskGameEvent":
            text = _format_game_event(entry)
        else:
            text = None
        if text:
            results.append(text)
    return results
